File: grasmere_routes/status.py
"""Customer status derivation. Mirrors the customer_status_v Postgres view
so we can apply the same rules in-memory during CSV import or tests."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date, datetime, timezone
from typing import Literal

CustomerStatus = Literal["live", "dormant", "no_history"]


@dataclass(frozen=True)
class StatusResult:
    status: CustomerStatus
    days_since_last_delivery: int | None


def derive_status(
    *,
    last_delivery_date: date | None,
    manually_confirmed_live_at: datetime | None,
    threshold_days: int = 180,
    today: date | None = None,
) -> StatusResult:
    today = today or date.today()
    cutoff = today.toordinal() - threshold_days

    if manually_confirmed_live_at is not None:
        # Strip tz for comparison; treat naive as UTC
        confirm_date = (
            manually_confirmed_live_at.date()
            if manually_confirmed_live_at.tzinfo is None
            else manually_confirmed_live_at.astimezone(timezone.utc).date()
        )
        if confirm_date.toordinal() >= cutoff:
            days_since = (
                (today - last_delivery_date).days if last_delivery_date else None
            )
            return StatusResult("live", days_since)

    if last_delivery_date is None:
        return StatusResult("no_history", None)

    days = (today - last_delivery_date).days
    return StatusResult("live" if days <= threshold_days else "dormant", days)

File: apps/web/lib/status.ts
/**
 * Customer status derivation. Logic mirrors the Postgres view
 * `customer_status_v` but is exposed for in-memory use during CSV import,
 * planning, and tests where we don't want to hit the DB.
 */

export type CustomerStatus = "live" | "dormant" | "no_history";

export interface StatusInputs {
  lastDeliveryDate: Date | null;
  manuallyConfirmedLiveAt: Date | null;
  /** Default 180 — keep aligned with config.dormancy_threshold_days */
  thresholdDays: number;
  /** Defaults to "now" in the caller's clock — testable */
  today?: Date;
}

const MS_PER_DAY = 24 * 60 * 60 * 1000;

export function deriveStatus(args: StatusInputs): {
  status: CustomerStatus;
  daysSinceLastDelivery: number | null;
} {
  const today = args.today ?? new Date();
  const cutoff = new Date(today.getTime() - args.thresholdDays * MS_PER_DAY);

  if (args.manuallyConfirmedLiveAt && args.manuallyConfirmedLiveAt >= cutoff) {
    return {
      status: "live",
      daysSinceLastDelivery: args.lastDeliveryDate
        ? Math.floor((today.getTime() - args.lastDeliveryDate.getTime()) / MS_PER_DAY)
        : null,
    };
  }

  if (!args.lastDeliveryDate) {
    return { status: "no_history", daysSinceLastDelivery: null };
  }

  const days = Math.floor((today.getTime() - args.lastDeliveryDate.getTime()) / MS_PER_DAY);
  return {
    status: days <= args.thresholdDays ? "live" : "dormant",
    daysSinceLastDelivery: days,
  };
}
